core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -499,6 +501,40 @@ protected long snapshotId() {
     return snapshotId;
   }
 
+  protected static <F extends ContentFile<F>> List<ManifestFile> writeFilesToManifests(
+      Iterable<F> files,
+      Supplier<ManifestWriter<F>> createWriter,
+      Long newFilesSequenceNumber,
+      long targetSizeBytes)
+      throws IOException {
+    List<ManifestFile> result = Lists.newArrayList();
+    Iterator<F> fileIterator = files.iterator();
+    ManifestWriter<F> writer = null;
```

Comment:
I think the writer needs to be wrapped in a try with resources to guarantee it gets closed. Also a nit which will be solved with try with resources: Could we initialize the writer upfront?

Reply:
Updated with try-finally, while try-with-resources seems not easy to implement.

> Could we initialize the writer upfront?

There are some UTs failed if do that. I tried the following:
```java
    try {
      writer = createWriter.get();
      while (fileIterator.hasNext()) {
        if (writer.length() >= targetSizeBytes) { 
          // here could produce an empty file because the intialize size of the newly writer could be larger than targetSizeBytes
          writer.close();
          result.add(writer.toManifestFile());
          writer = createWriter.get();
        }

        F file = fileIterator.next();
        if (newFilesSequenceNumber == null) {
          writer.add(file);
        } else {
          writer.add(file, newFilesSequenceNumber);
        }
      }
    } finally {
      if (writer != null) {
        writer.close();
      }
    }
```